README.md
# tfwrapper

Build of master branch: [![TravisCI](https://api.travis-ci.org/manheim/tfwrapper.svg?branch=master)](https://travis-ci.org/manheim/tfwrapper)

Documentation: [http://www.rubydoc.info/gems/tfwrapper/](http://www.rubydoc.info/gems/tfwrapper/)

tfwrapper provides Rake tasks for working with [Hashicorp Terraform](https://www.terraform.io/) 0.9 through 0.11, ensuring proper initialization and passing in variables from the environment or Ruby, as well as optionally pushing some information to Consul. tfwrapper also attempts to detect and retry
failed runs due to AWS throttling or access denied errors.

## Overview

This Gem provides the following Rake tasks:

* __tf:init__ - run ``terraform init`` to pull down dependency modules and configure remote
  state backend. This task also checks that any configured environment variables are set and
  that the ``terraform`` version is compatible with this gem.
* __tf:plan__ - run ``terraform plan`` with all variables and configuration, and TF variables written to disk. You can specify
  one or more optional [resource address](https://www.terraform.io/docs/internals/resource-addressing.html) targets to pass to
  terraform with the ``-target`` flag as Rake task arguments, i.e. ``bundle exec rake tf:plan[aws_instance.foo[1]]`` or
  ``bundle exec rake tf:plan[aws_instance.foo[1],aws_instance.bar[2]]``; see the
  [plan documentation](https://www.terraform.io/docs/commands/plan.html) for more information. By default this will use
  [terraform_landscape](https://github.com/coinbase/terraform-landscape/blob/master/README.md) for formatting the plan
  output if the ``terraform_landscape`` gem is installed; see the [section below](#terraform-landscape) for more information.
* __tf:apply__ - run ``terraform apply`` with all variables and configuration, and TF variables written to disk. You can specify
  one or more optional [resource address](https://www.terraform.io/docs/internals/resource-addressing.html) targets to pass to
  terraform with the ``-target`` flag as Rake task arguments, i.e. ``bundle exec rake tf:apply[aws_instance.foo[1]]`` or
  ``bundle exec rake tf:apply[aws_instance.foo[1],aws_instance.bar[2]]``; see the
  [apply documentation](https://www.terraform.io/docs/commands/apply.html) for more information. This also runs a plan first.
* __tf:refresh__ - run ``terraform refresh``
* __tf:destroy__ - run ``terraform destroy`` with all variables and configuration, and TF variables written to disk. You can specify
  one or more optional [resource address](https://www.terraform.io/docs/internals/resource-addressing.html) targets to pass to
  terraform with the ``-target`` flag as Rake task arguments, i.e. ``bundle exec rake tf:destroy[aws_instance.foo[1]]`` or
  ``bundle exec rake tf:destroy[aws_instance.foo[1],aws_instance.bar[2]]``; see the
  [destroy documentation](https://www.terraform.io/docs/commands/destroy.html) for more information.
* __tf:write_tf_vars__ - used as a prerequisite for other tasks; write Terraform variables to file on disk
* __tf:output__ - run ``terraform output``
* __tf:output_json__ - run ``terraform output -json``

## Installation

__Note:__ tfwrapper only supports Ruby >= 2.0.0. The effort to maintain compatibility
with 1.9.3 is simply too high to justify.

Add to your ``Gemfile``:

```ruby
gem 'tfwrapper', '~> 0.6.1'
```

### Supported Terraform Versions

tfwrapper only supports terraform 0.9-0.11. It is tested against multiple versions from 0.9.2 to 0.11.14.

### Terraform Landscape

The [terraform_landscape](https://github.com/coinbase/terraform-landscape/blob/master/README.md) gem provides enhanced formatting of ``terraform plan`` output including colorization of changes and human-friendly diffs (i.e. diffs of JSON rendered with pretty-printing). By default ``plan`` output will be passed through terraform_landscape if the ``terraform_landscape`` gem is available. To enable this, add ``gem 'terraform_landscape', '~> 0.1.17'`` to your ``Gemfile``. __Note__ that we rely on an undocumented internal API of terraform_landscape to achieve this; the formatting code will fall back to unformatted output in case of an error.

If you wish to disable terraform_landscape output even when the gem is installed, pass ``disable_landscape: true`` as an option to ``install_tasks()``:

```ruby
TFWrapper::RakeTasks.install_tasks('.', disable_landscape: true)
```

In previous versions or when terraform_landscape is not installed, the output of all terraform commands is streamed in realtime. Since terraform_landscape requires the full and complete ``plan`` output in order to reformat it, this is no longer the case. By default when terraform_landscape is installed and not disabled the ``plan`` task will not produce any output until complete, at which point it will print the landscape-formatted output all at once. This behavior can be controlled with the ``:landscape_progress`` option of ``install_tasks()``, which takes one of the following values:

* ``nil``, the default, to not produce any output until the command is complete at which point the landscape-formatted output will be shown.
* ``:dots`` to print a dot to STDOUT for every line of ``terraform plan`` output and then print the landscape-formatted output when complete.
* ``:lines`` to print a dot followed by a newline to STDOUT for every line of ``terraform plan`` output and then print the landscape-formatted output when complete. This is useful for systems like Jenkins that line-buffer output (and don't display anything until a newline is encountered).
* ``:stream`` to stream the ``terraform plan`` output in realtime (as was the previous behavior) and then print the landscape-formatted output when complete. Note that this will result in the output containing the complete unformatted ``terraform plan`` output, followed by the landscape-formatted output.

## Usage

To use the Terraform rake tasks, require the module in your Rakefile and use the
``install_tasks`` method to set up the tasks. ``install_tasks`` takes one mandatory parameter,
``tf_dir`` specifying the relative path (from the Rakefile) to the Terraform configuration.

For a directory layout like:

```
.
├── bar.tf
├── foo.tf
├── main.tf
└── Rakefile
```

The minimal ``Rakefile`` would be:

```ruby
require 'tfwrapper/raketasks'

TFWrapper::RakeTasks.install_tasks('.')
```

``rake -T`` output:

```
rake tf:apply[target]        # Apply a terraform plan that will provision your resources; specify optional CSV targets
rake tf:destroy[target]      # Destroy any live resources that are tracked by your state files; specify optional CSV targets
rake tf:init                 # Run terraform init with appropriate arguments
rake tf:plan[target]         # Output the set plan to be executed by apply; specify optional CSV targets
rake tf:write_tf_vars        # Write PWD/build.tfvars.json
```

You can also point ``tf_dir`` to an arbitrary directory relative to the Rakefile, such as when your terraform
configurations are nested below the Rakefile:

```
.
├── infrastructure
│   └── terraform
│       ├── bar.tf
│       ├── foo.tf
│       └── main.tf
├── lib
├── Rakefile
└── spec
```

Rakefile:

```ruby
require 'tfwrapper/raketasks'

TFWrapper::RakeTasks.install_tasks('infrastructure/terraform')
```

### Environment Variables to Terraform Variables

If you wish to bind the values of environment variables to Terraform variables, you can specify a mapping
of Terraform variable name to environment variable name in the ``tf_vars_from_env`` option; these variables
will be automatically read from the environment and passed into Terraform with the appropriate names. The following
example sets the ``consul_address`` terraform variable to the value of the ``CONSUL_HOST`` environment variable
(defaulting it to ``consul.example.com:8500`` if it is not already set in the environment),
and likewise for the ``environment`` terraform variable from the ``ENVIRONMENT`` env var.

```ruby
require 'tfwrapper/raketasks'
ENV['CONSUL_HOST'] ||= 'consul.example.com:8500'

TFWrapper::RakeTasks.install_tasks(
  '.',
  tf_vars_from_env: {
    'consul_address'           => 'CONSUL_HOST',
    'environment'              => 'ENVIRONMENT',
  }
)
```

These variables are tested to be set in the environment with a non-empty value, and will raise an error if any are
missing or empty. If some should be allowed to be missing or empty empty, pass a ``allowed_empty_vars`` list with
their environment variable names.

### Ruby Variables to Terraform Variables

If you wish to explicitly bind values from your Ruby code to terraform variables, you can do this with
the ``tf_extra_vars`` option. Variables specified in this way will override same-named variables populated
via ``tf_vars_from_env``. In the following example, the ``foobar`` terraform variable will have a value
of ``baz``, regardless of what the ``FOOBAR`` environment variable is set to, and the ``hostname``
terraform variable will be set to the hostname (``Socket.gethostname``) of the system Rake is running on:

```ruby
require 'socket'
require 'tfwrapper/raketasks'

ENV['FOOBAR'] ||= 'not_baz'

TFWrapper::RakeTasks.install_tasks(
  '.',
  tf_vars_from_env: {
    'foobar' => 'FOOBAR'
  },
  tf_extra_vars: {
    'foobar'   => 'baz',
    'hostname' => Socket.gethostname
  }
)
```

### Namespace Prefixes for Multiple Configurations

If you need to work with multiple different Terraform configurations, this is possible
by adding a namespace prefix and calling ``install_tasks`` multiple times. The following example
will produce two sets of terraform Rake tasks; one with the default ``tf:`` namespace
that acts on the configurations under ``tf/foo``, and one with a ``bar_tf:`` namespace
that acts on the configurations under ``tf/bar``. You can use as many namespaces as
you want.

Directory tree:

```
.
├── Rakefile
└── tf
    ├── bar
    │   └── bar.tf
    └── foo
        └── foo.tf
```

Rakefile:

```ruby
require 'tfwrapper/raketasks'

# foo/ (default) terraform tasks
TFWrapper::RakeTasks.install_tasks('tf/foo')

# bar/ terraform tasks
TFWrapper::RakeTasks.install_tasks('tf/bar', namespace_prefix: 'bar')
```

``rake -T`` output:

```
rake bar_tf:apply[target]    # Apply a terraform plan that will provision your resources; specify optional CSV targets
rake bar_tf:destroy[target]  # Destroy any live resources that are tracked by your state files; specify optional CSV targets
rake bar_tf:init             # Run terraform init with appropriate arguments
rake bar_tf:plan[target]     # Output the set plan to be executed by apply; specify optional CSV targets
rake bar_tf:write_tf_vars    # Write PWD/bar_build.tfvars.json
rake tf:apply[target]        # Apply a terraform plan that will provision your resources; specify optional CSV targets
rake tf:destroy[target]      # Destroy any live resources that are tracked by your state files; specify optional CSV targets
rake tf:init                 # Run terraform init with appropriate arguments
rake tf:plan[target]         # Output the set plan to be executed by apply; specify optional CSV targets
rake tf:write_tf_vars        # Write PWD/build.tfvars.json
```

### Backend Configuration Options

``install_tasks`` accepts a ``backend_config`` hash of options to pass as backend configuration
to ``terraform init`` via the ``-backend-config='key=value'`` command line argument. This can
be used when you need to pass some backend configuration in from the environment, such as a
specific remote state storage path, credentials, etc.

For a simple example, assume we aren't using [state environments](https://www.terraform.io/docs/state/environments.html)
but instead opt to use specific paths based on a ``ENVIRONMENT`` environment variable.

Our terraform configuration might include something like:

```
terraform {
  required_version = "> 0.9.0"
  backend "consul" {
    address = "consul.example.com:8500"
  }
}

variable "environment" {}
```

And the Rakefile would pass in the path to store state in Consul, as well as
passing the ``ENVIRONMENT`` env var into Terraform for use:

```ruby
require 'tfwrapper/raketasks'

TFWrapper::RakeTasks.install_tasks(
  '.',
  tf_vars_from_env: {'environment' => 'ENVIRONMENT'},
  backend_config: {'path' => "terraform/foo/#{ENVIRONMENT}"}
)
```

### Calling Procs At Beginning and End of Tasks

Version 0.4.0 of tfwrapper introduced the ability to call arbitrary Ruby Procs from within the Rake tasks,
at the beginning and end of the task (i.e. before and after the terraform-handling code within the task).
This is accomplished via the ``:before_proc`` and ``:after_proc`` options, each taking a Proc instance.

The Procs take two positional arguments; a ``String`` containing the full, namespaced name of the Rake task
it was called from, and the ``String`` ``tf_dir`` argument passed to the TFWrapper::RakeTasks class (exactly as specified).

This could be used for things such as showing the output of state after a terraform run,
triggering a [tfenv](https://github.com/kamatama41/tfenv) installation, etc.

```ruby
require 'tfwrapper/raketasks'

TFWrapper::RakeTasks.install_tasks(
  '.',
  before_proc: Proc.new do |taskname, tfdir|
    next unless taskname == 'tf:apply' # example of only executing for apply task in default namespace
    puts "Executing #{taskname} task with tfdir=#{tfdir}"
  end,
  after_proc: Proc.new do |taskname, tfdir|
    puts "Executed #{taskname} task with tfdir=#{tfdir}"
  end
)
```

### Environment Variables to Consul

tfwrapper also includes functionality to push environment variables to Consul
(as a JSON object) after a successful apply. This is mainly useful when running

ChangeLog.md
Version 0.6.3

  - Fix bug with ``tf_sensitive_vars``. Prevent values from being printed in ``update_consul_stack_env_vars``.

Version 0.6.2

  - Add ``tf_sensitive_vars`` option.

Version 0.6.1

  - Add ``allowed_empty_vars`` option.
  - Only support terraform up to 0.11.14.

Version 0.6.0

  - Include full terraform output when a terraform run fails with output including ``hrottling``, ``status code: 403``, or ``status code: 401``. Previously terraform output was suppressed in these cases, which causes confusion with providers other than "aws" that include these strings in their failure output.
  - In ``Gemfile``, for testing, pin terraform-landscape gem to 0.2.2 for Ruby 2.x compatibility.
  - Pin ``rb-inotify`` dependency to 0.9.10 to allow build with Ruby < 2.2
  - Switch testing from circleci.com to travis-ci.org
  - Pin terraform_landscape dependency to 0.2.2 for compatibility with ruby < 2.5
  - Acceptance tests:
    - Use HashiCorp checkpoint API instead of GitHub Releases API to find latest terraform version, to work around query failures
    - Pin consul provider versions to 1.0.0 to fix intermittent failures
  - Switch Ruby versions used in tests to latest TravisCI versions
  - Stop acceptance testing terraform 0.9.x, as newer versions require pinning the consul provider version but 0.9 doesn't support versioned providers.

Version 0.5.1

  - Fix bug where terraform plan errors were suppressed if a plan run with landscape support enabled exited non-zero.

Version 0.5.0

  - Add support for using terraform_landscape gem, if present, to reformat plan output; see README for usage.
  - Add CircleCI testing under ruby 2.4.1, and acceptance tests for terraform 0.11.2.

Version 0.4.1

  - Upgrade rubocop, yard and diplomat development dependency versions
  - Pin `cri` development dependency to 2.9 to retain ruby 2.0 support

Version 0.4.0

  - Add support for calling Procs at the beginning and end of each task.
  - Documentation cleanup.

Version 0.3.0

  - Add `tf:output` and `tf:output_json` Rake tasks
  - Have TFWrapper::RakeTasks store the current terraform version as an instance variable;
    also make this accessible as tf_version.
  - Add '-auto-approve' to terraform apply command if running with tf >= 0.10.0
  - Update acceptance tests to work with terraform 0.9.x and 0.10.x
  - Acceptance test logic to find the latest terraform release version from GitHub API
  - Run acceptance tests for multiple terraform versions (currently: 0.9.2, 0.9.7, 0.10.0, 0.10.2, latest)

Version 0.2.0

  - initial release (migrated from previous internal/private gem)

Version 0.1.0

  - no working public release


lib/tfwrapper/raketasks.rb
# frozen_string_literal: true

require 'tfwrapper/helpers'
require 'json'
require 'rake'
require 'rubygems'
require 'tfwrapper/version'

# :nocov:
begin
  require 'terraform_landscape'
  HAVE_LANDSCAPE = true
rescue LoadError
  HAVE_LANDSCAPE = false
end
# :nocov:

module TFWrapper
  # Generates Rake tasks for working with Terraform.
  class RakeTasks
    include Rake::DSL if defined? Rake::DSL

    class << self
      # set when installed
      attr_accessor :instance

      # Install the Rake tasks for working with Terraform. For full parameter
      # documentation, see {TFWrapper::RakeTasks#initialize}
      #
      # @param (see #initialize)
      def install_tasks(tf_dir, opts = {})
        new(tf_dir, opts).install
      end
    end

    def min_tf_version
      Gem::Version.new('0.9.0')
    end

    attr_reader :tf_version

    # Generate Rake tasks for working with Terraform.
    #
    # @param tf_dir [String] Terraform config directory, relative to Rakefile.
    #   Set to '.' if the Rakefile is in the same directory as the ``.tf``
    #   configuration files.
    # @option opts [Hash] :backend_config hash of Terraform remote state
    #   backend configuration options, to override or supplement those in
    #   the terraform configuration. See the
    #   [Remote State](https://www.terraform.io/docs/state/remote.html)
    #   documentation for further information.
    # @option opts [String] :namespace_prefix if specified and not nil, this
    #   will put all tasks in a "#{namespace_prefix}_tf:" namespace instead
    #   of "tf:". This allows using manheim_helpers for multiple terraform
    #   configurations in the same Rakefile.
    # @option opts [Hash] :tf_vars_from_env hash of Terraform variables to the
    #   (required) environment variables to populate their values from
    # @option opts [Hash] :allowed_empty_vars array of environment variable
    #   names (specified in :tf_vars_from_env) to allow to be empty or missing.
    # @option opts [Hash] :tf_extra_vars hash of Terraform variables to their
    #   values; overrides any same-named keys in ``tf_vars_from_env``
    # @option opts [Array] :tf_sensitive_vars list of Terraform variables
    #   which should not be printed
    # @option opts [String] :consul_url URL to access Consul at, for the
    #   ``:consul_env_vars_prefix`` option.
    # @option opts [String] :consul_env_vars_prefix if specified and not nil,
    #   write the environment variables used from ``tf_vars_from_env``
    #   and their values to JSON at this path in Consul. This should have
    #   the same naming constraints as ``consul_prefix``.
    # @option opts [Proc] :before_proc Proc instance to call before executing
    #   the body of each task. Called with two arguments, the String full
    #   (namespaced) name of the task being executed, and ``tf_dir``. Returning
    #   or breaking from this Proc will cause the task to not execute; to exit
    #   the Proc early, use ``next``.
    # @option opts [Proc] :after_proc Proc instance to call after executing
    #   the body of each task. Called with two arguments, the String full
    #   (namespaced) name of the task being executed, and ``tf_dir``. This will
    #   not execute if the body of the task fails.
    # @option opts [Bool] :disable_landscape By default, if the
    #  ``terraform_landscape`` gem can be loaded, it will be used to reformat
    #  the output of ``terraform plan``. If this is not desired, set to
    #  ``true`` to disable landscale. Default: ``false``.
    # @option opts [Symbol, nil] :landscape_progress The ``terraform_landscape``
    #  code used to reformat plan output requires the full output of the
    #  complete ``plan`` execution. By default, this means that when landscape
    #  is used, no output will appear from the time ``terraform plan`` begins
    #  until the command is complete. If progress output is desired, this option
    #  can be set to one of the following: ``:dots`` to print a dot to STDOUT
    #  for every line of ``terraform plan`` output, ``:lines`` to print a dot
    #  followed by a newline (e.g. for systems like Jenkins that line buffer)
    #  for every line of ``terraform plan`` output, or ``:stream`` to stream
    #  the raw ``terraform plan`` output (which will then be followed by the
    #  reformatted landscape output). Default is ``nil`` to show no progress
    #  output.
    def initialize(tf_dir, opts = {})
      # find the directory that contains the Rakefile
      rakedir = File.realpath(Rake.application.rakefile)
      rakedir = File.dirname(rakedir) if File.file?(rakedir)
      @tf_dir = File.realpath(File.join(rakedir, tf_dir))
      @ns_prefix = opts.fetch(:namespace_prefix, nil)
      @consul_env_vars_prefix = opts.fetch(:consul_env_vars_prefix, nil)
      @tf_vars_from_env = opts.fetch(:tf_vars_from_env, {})
      @allowed_empty_vars = opts.fetch(:allowed_empty_vars, [])
      @tf_sensitive_vars = opts.fetch(:tf_sensitive_vars, [])
      @tf_extra_vars = opts.fetch(:tf_extra_vars, {})
      @backend_config = opts.fetch(:backend_config, {})
      @consul_url = opts.fetch(:consul_url, nil)
      @disable_landscape = opts.fetch(:disable_landscape, false)
      @landscape_progress = opts.fetch(:landscape_progress, nil)
      unless [:dots, :lines, :stream, nil].include?(@landscape_progress)
        raise(
          ArgumentError,
          'landscape_progress option must be one of: ' \
          '[:dots, :lines, :stream, nil]'
        )
      end
      @before_proc = opts.fetch(:before_proc, nil)
      if !@before_proc.nil? && !@before_proc.is_a?(Proc)
        raise(
          TypeError,
          'TFWrapper::RakeTasks.initialize option :before_proc must be a ' \
          'Proc instance, not a ' + @before_proc.class.name
        )
      end
      @after_proc = opts.fetch(:after_proc, nil)
      if !@after_proc.nil? && !@after_proc.is_a?(Proc)
        raise(
          TypeError,
          'TFWrapper::RakeTasks.initialize option :after_proc must be a Proc ' \
          'instance, not a ' + @after_proc.class.name
        )
      end
      # default to lowest possible version; this is set in the 'init' task
      @tf_version = Gem::Version.new('0.0.0')
      # rubocop:disable Style/GuardClause
      if @consul_url.nil? && !@consul_env_vars_prefix.nil?
        raise StandardError, 'Cannot set env vars in Consul when consul_url ' \
          'option is nil.'
      end
      # rubocop:enable Style/GuardClause
    end

    def nsprefix
      if @ns_prefix.nil?
        'tf'.to_sym
      else
        "#{@ns_prefix}_tf".to_sym
      end
    end

    # install all Rake tasks - calls other install_* methods
    # rubocop:disable Metrics/CyclomaticComplexity
    def install
      install_init
      install_plan
      install_apply
      install_refresh
      install_destroy
      install_write_tf_vars
      install_output
    end

    # add the 'tf:init' Rake task. This checks environment variables,
    # runs ``terraform -version``, and then runs ``terraform init`` with
    # the ``backend_config`` options, if any.
    def install_init
      namespace nsprefix do
        desc 'Run terraform init with appropriate arguments'
        task :init do |t|
          @before_proc.call(t.name, @tf_dir) unless @before_proc.nil?
          TFWrapper::Helpers.check_env_vars(
            @tf_vars_from_env.values, @allowed_empty_vars
          )
          @tf_version = check_tf_version
          cmd = [
            'terraform',
            'init',
            '-input=false'
          ].join(' ')
          @backend_config.each do |k, v|
            cmd = cmd + ' ' + "-backend-config='#{k}=#{v}'"
          end
          terraform_runner(cmd)
          @after_proc.call(t.name, @tf_dir) unless @after_proc.nil?
        end
      end
    end

    # add the 'tf:plan' Rake task
    def install_plan
      namespace nsprefix do
        desc 'Output the set plan to be executed by apply; specify ' \
          'optional CSV targets'
        task :plan, [:target] => [
          :"#{nsprefix}:init",
          :"#{nsprefix}:write_tf_vars"
        ] do |t, args|
          @before_proc.call(t.name, @tf_dir) unless @before_proc.nil?
          cmd = cmd_with_targets(
            ['terraform', 'plan', "-var-file #{var_file_path}"],
            args[:target],
            args.extras
          )

          stream_type = if HAVE_LANDSCAPE && !@disable_landscape
                          @landscape_progress
                        else
                          :stream
                        end
          outerr = terraform_runner(cmd, progress: stream_type)
          landscape_format(outerr) if HAVE_LANDSCAPE && !@disable_landscape
          @after_proc.call(t.name, @tf_dir) unless @after_proc.nil?
        end
      end
    end

    # add the 'tf:apply' Rake task
    def install_apply
      namespace nsprefix do
        desc 'Apply a terraform plan that will provision your resources; ' \
          'specify optional CSV targets'
        task :apply, [:target] => [
          :"#{nsprefix}:init",
          :"#{nsprefix}:write_tf_vars",
          :"#{nsprefix}:plan"
        ] do |t, args|
          @before_proc.call(t.name, @tf_dir) unless @before_proc.nil?
          cmd_arr = %w[terraform apply]
          cmd_arr << '-auto-approve' if tf_version >= Gem::Version.new('0.10.0')
          cmd_arr << "-var-file #{var_file_path}"
          cmd = cmd_with_targets(
            cmd_arr,
            args[:target],
            args.extras
          )
          terraform_runner(cmd)

          update_consul_stack_env_vars unless @consul_env_vars_prefix.nil?
          @after_proc.call(t.name, @tf_dir) unless @after_proc.nil?
        end
      end
    end

    # add the 'tf:refresh' Rake task
    def install_refresh
      namespace nsprefix do
        task refresh: [
          :"#{nsprefix}:init",
          :"#{nsprefix}:write_tf_vars"
        ] do |t|
          @before_proc.call(t.name, @tf_dir) unless @before_proc.nil?
          cmd = [
            'terraform',
            'refresh',
            "-var-file #{var_file_path}"
          ].join(' ')

          terraform_runner(cmd)
          @after_proc.call(t.name, @tf_dir) unless @after_proc.nil?
        end
      end
    end

    # add the 'tf:output' Rake task
    def install_output
      namespace nsprefix do
        task output: [
          :"#{nsprefix}:init",
          :"#{nsprefix}:refresh"
        ] do |t|
          @before_proc.call(t.name, @tf_dir) unless @before_proc.nil?
          terraform_runner('terraform output')
          @after_proc.call(t.name, @tf_dir) unless @after_proc.nil?
        end
        task output_json: [
          :"#{nsprefix}:init",
          :"#{nsprefix}:refresh"
        ] do |t|
          @before_proc.call(t.name, @tf_dir) unless @before_proc.nil?
          terraform_runner('terraform output -json')
          @after_proc.call(t.name, @tf_dir) unless @after_proc.nil?
        end
      end
    end

    # add the 'tf:destroy' Rake task
    def install_destroy
      namespace nsprefix do
        desc 'Destroy any live resources that are tracked by your state ' \
          'files; specify optional CSV targets'
        task :destroy, [:target] => [
          :"#{nsprefix}:init",
          :"#{nsprefix}:write_tf_vars"
        ] do |t, args|
          @before_proc.call(t.name, @tf_dir) unless @before_proc.nil?
          cmd = cmd_with_targets(
            ['terraform', 'destroy', '-force', "-var-file #{var_file_path}"],
            args[:target],
            args.extras
          )

          terraform_runner(cmd)
          @after_proc.call(t.name, @tf_dir) unless @after_proc.nil?
        end
      end
    end

    def var_file_path
      if @ns_prefix.nil?
        File.absolute_path('build.tfvars.json')
      else
        File.absolute_path("#{@ns_prefix}_build.tfvars.json")
      end
    end

    # add the 'tf:write_tf_vars' Rake task
    def install_write_tf_vars
      namespace nsprefix do
        desc "Write #{var_file_path}"
        task :write_tf_vars do |t|
          @before_proc.call(t.name, @tf_dir) unless @before_proc.nil?
          tf_vars = terraform_vars
          puts 'Terraform vars:'
          tf_vars.sort.map do |k, v|
            redacted_list = (%w[aws_access_key aws_secret_key] +
                             @tf_sensitive_vars)
            if redacted_list.include?(k)
              puts "#{k} => (redacted)"
            else
              puts "#{k} => #{v}"
            end
          end
          File.open(var_file_path, 'w') do |f|
            f.write(tf_vars.to_json)
          end
          STDERR.puts "Terraform vars written to: #{var_file_path}"
          @after_proc.call(t.name, @tf_dir) unless @after_proc.nil?
        end
      end
    end

    def terraform_vars
      res = {}
      @tf_vars_from_env.each { |tfname, envname| res[tfname] = ENV[envname] }
      @tf_extra_vars.each { |name, val| res[name] = val }
      res
    end

    # Given a string of terraform plan output, format it with
    # terraform_landscape and print the result to STDOUT.
    def landscape_format(output)
      p = TerraformLandscape::Printer.new(
        TerraformLandscape::Output.new(STDOUT)
      )
      p.process_string(output)
    rescue StandardError, ScriptError => ex
      STDERR.puts 'Exception calling terraform_landscape to reformat ' \
                  "output: #{ex.class.name}: #{ex}"
      puts output unless @landscape_progress == :stream
    end

    # Run a Terraform command, providing some useful output and handling AWS
    # API rate limiting gracefully. Raises StandardError on failure. The command
    # is run in @tf_dir.
    #
    # @param cmd [String] Terraform command to run
    # @option opts [Hash] :progress How to handle streaming output. Possible
    #  values are ``:stream`` (default) to stream each line in STDOUT/STDERR
    #  to STDOUT, ``:dots`` to print a dot for each line, ``:lines`` to print
    #  a dot followed by a newline for each line, or ``nil`` to not stream any
    #  output at all.
    # @return [String] combined STDOUT and STDERR
    # rubocop:disable Metrics/PerceivedComplexity
    def terraform_runner(cmd, opts = {})
      stream_type = opts.fetch(:progress, :stream)
      unless [:dots, :lines, :stream, nil].include?(stream_type)
        raise(
          ArgumentError,
          'progress option must be one of: [:dots, :lines, :stream, nil]'
        )
      end
      require 'retries'
      STDERR.puts "terraform_runner command: '#{cmd}' (in #{@tf_dir})"
      out_err = nil
      status = nil
      # exponential backoff as long as we're getting 403s
      handler = proc do |exception, attempt_number, total_delay|
        STDERR.puts "terraform_runner failed with #{exception}; retry " \
          "attempt #{attempt_number}; #{total_delay} seconds have passed."
      end
      status = -1
      with_retries(
        max_tries: 5,
        handler: handler,
        base_sleep_seconds: 1.0,
        max_sleep_seconds: 10.0
      ) do
        # this streams STDOUT and STDERR as a combined stream,
        # and also captures them as a combined string
        out_err, status = TFWrapper::Helpers.run_cmd_stream_output(
          cmd, @tf_dir, progress: stream_type
        )
        if status != 0 && out_err.include?('hrottling')
          raise StandardError, "#{out_err}\nTerraform hit AWS API rate limiting"
        end
        if status != 0 && out_err.include?('status code: 403')
          raise StandardError, "#{out_err}\nTerraform command got 403 error " \
            '- access denied or credentials not propagated'
        end
        if status != 0 && out_err.include?('status code: 401')
          raise StandardError, "#{out_err}\nTerraform command got 401 error " \
            '- access denied or credentials not propagated'
        end
      end
      # end exponential backoff
      unless status.zero?
        # if we weren't streaming output, send it now
        STDERR.puts out_err unless stream_type == :stream
        raise StandardError, "Errors have occurred executing: '#{cmd}' " \
          "(exited #{status})"
      end
      STDERR.puts "terraform_runner command '#{cmd}' finished and exited 0"
      out_err
    end
    # rubocop:enable Metrics/PerceivedComplexity

    # Check that the terraform version is compatible
    def check_tf_version
      # run: terraform -version
      all_out_err, exit_status = TFWrapper::Helpers.run_cmd_stream_output(
        'terraform version', @tf_dir
      )
      unless exit_status.zero?
        raise StandardError, "ERROR: 'terraform -version' exited " \
          "#{exit_status}: #{all_out_err}"
      end
      all_out_err = all_out_err.strip
      # Find the terraform version string
      m = /Terraform v(\d+\.\d+\.\d+).*/.match(all_out_err)
      unless m
        raise StandardError, 'ERROR: could not determine terraform version ' \
          "from 'terraform -version' output: #{all_out_err}"
      end
      # the version will be a string like:
      # Terraform v0.9.2
      # or:
      # Terraform v0.9.3-dev (<GIT SHA><+CHANGES>)
      tf_ver = Gem::Version.new(m[1])
      unless tf_ver >= min_tf_version
        raise StandardError, "ERROR: tfwrapper #{TFWrapper::VERSION} is only " \
          "compatible with Terraform >= #{min_tf_version} but your terraform " \
          "binary reports itself as #{m[1]} (#{all_out_err})"
      end
      puts "Running with: #{all_out_err}"
      tf_ver
    end

    # update stack status in Consul
    def update_consul_stack_env_vars
      require 'diplomat'
      require 'json'
      data = {}
      @tf_vars_from_env.each_value { |k| data[k] = ENV[k] }

      Diplomat.configure do |config|
        config.url = @consul_url
      end

      puts "Writing stack information to #{@consul_url} at: "\
        "#{@consul_env_vars_prefix}"

      redacted_list = (%w[aws_access_key aws_secret_key] +
        @tf_sensitive_vars)
      sanitized_data = data.clone
      @tf_vars_from_env.each do |k, v|
        # We are trying to determine which ENV var maps
        #   to the sensitive terraform variable
        sanitized_data[v] = '(redacted)' if redacted_list.include?(k)
      end
      puts JSON.pretty_generate(sanitized_data)
      raw = JSON.generate(data)
      Diplomat::Kv.put(@consul_env_vars_prefix, raw)
    end

    # Create a terraform command line with optional targets specified; targets
    # are inserted between cmd_array and suffix_array.
    #
    # This is intended to simplify parsing Rake task arguments and inserting
    # them into the command as targets; to get a Rake task to take a variable
    # number of arguments, we define a first argument (``:target``) which is
    # either a String or nil. Any additional arguments specified end up in
    # ``args.extras``, which is either nil or an Array of additional String
    # arguments.
    #
    # @param cmd_array [Array] array of the beginning parts of the terraform
    #   command; usually something like:
    #   ['terraform', 'ACTION', '-var'file', 'VAR_FILE_PATH']
    # @param target [String] the first target parameter given to the Rake
    #   task, or nil.
    # @param extras [Array] array of additional target parameters given to the
    #   Rake task, or nil.
    def cmd_with_targets(cmd_array, target, extras)
      final_arr = cmd_array
      final_arr.concat(['-target', target]) unless target.nil?
      # rubocop:disable Style/SafeNavigation
      extras.each { |e| final_arr.concat(['-target', e]) } unless extras.nil?
      # rubocop:enable Style/SafeNavigation
      final_arr.join(' ')
    end
  end
end
